diff --git a/source/i18n/unicode/numberrangeformatter.h b/source/i18n/unicode/numberrangeformatter.h
index 8ca20f31..64e43bc9 100644
--- a/source/i18n/unicode/numberrangeformatter.h
+++ b/source/i18n/unicode/numberrangeformatter.h
@@ -68,6 +68,9 @@ struct UFormattedNumberRangeImpl;
 
 } // namespace impl
 
+
+#if 0
+
 /**
  * \cond
  * Export an explicit template instantiation. See datefmt.h
@@ -84,6 +87,8 @@ namespace number {  // icu::number
 #endif
 /** \endcond */
 
+#endif
+
 // Other helper classes would go here, but there are none.
 
 namespace impl {  // icu::number::impl
